pkg/cmd/pulumi/search.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	envutil "github.com/pulumi/pulumi/sdk/v3/go/common/util/env"
+	"github.com/spf13/cobra"
+)
+
+func newSearchCmd() *cobra.Command {
+	var orgName string
+	var queryParams *[]string
+	cmd := &cobra.Command{
+		Use:   "search",
+		Short: "Search for resources in Pulumi Cloud",
+		Long:  "Search for resources in Pulumi Cloud.",
+		Args:  cmdutil.NoArgs,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
```

Comment:
For command structure, we've been following a slightly different pattern for newer commands to make them more testable. Examples are in state_edit.go, whoami.go, state_upgrade.go, package_publish.go, stack_output.go, stack_init.go, but basically, we use `newFooCmd() *cobra.Command` as a thin integration point between cobra and the real implementation, which takes the form:

```go
type searchCmd struct {
  // flags go here
  orgName string
  queryParams []string

  // optional global state goes here
}

func (cmd *searchCmd) Run(ctx context.Context, args []string) error {
  // ...
}
```

So the integration point looks like this:

```go
func newSearchCmd() *cobra.Command {
  var scmd searchCmd
  cmd := &cobra.Command{
    // ...
    Run: cmdutil.RunFunc(... {
	ctx := commandContext()
	return scmd.Run(ctx, args)
    })
  }
  
  cmd.PersistentFlags().StringVarP(&scmd.orgName, ...)
  // other flags
  return cmd
}
```

This allows us to make global dependencies (`os.Getenv`, `os.Stdout`, some top-level functions, etc.) injectable as needed. e.g.,

```go
type searchCmd struct {
  // ...

  Getenv func(string) string // == os.Getenv
}

func (cmd *searchCmd) Run(...) error {
  if cmd.Getenv == nil {
    cmd.Getenv = os.Getenv
  }
  
  // ...
  
  if cmd.Getenv("FOO_IS_SET") != "" {
    // ...
  }
}
```

e.g., 'state upgrade' uses it to inject the currentBackend function:

https://github.com/pulumi/pulumi/blob/504abb2c681e0c3567d28d02911d07a731b324a7/pkg/cmd/pulumi/state_upgrade.go#L63-L65